crmprtd/insert.py

```diff
@@ -8,11 +8,217 @@
 speed and reliability. This phase is common to all networks.
 """
 
+from math import floor
+from sqlalchemy import and_
+import logging
+import time
+from sqlalchemy.exc import IntegrityError
+
+from crmprtd.db_exceptions import UniquenessError, InsertionError
+from pycds import Obs
+
+
+log = logging.getLogger('crmprtd.insert')
+
+
+class DBMetrics(object):
+    '''Keep track of database metrics during the insertion process.
+    '''
+
+    def __init__(self):
+        self.successes = 0
+        self.failures = 0
+
+    def clear(self):
+        self.successes = 0
+        self.failures = 0
+
+
+class Timer(object):
+    def __enter__(self):
+        self.start = time.time()
+        return self
+
+    def __exit__(self, *args):
+        self.end = time.time()
+        self.run_time = self.end - self.start
+
+
+def chunks(list, chunk_size):
+    for i in range(0, len(list), chunk_size):
+        yield list[i:i+chunk_size]
+
+
+def get_sample_indices(num_obs, sample_size):
+    if sample_size > num_obs:
+        sample_size = num_obs
+
+    index_buffer = floor(num_obs/sample_size)
+
+    sample_indices = [0]
+    index = 0
+    for i in range(sample_size - 1):
+        index += index_buffer
+        sample_indices.append(index)
+
+    return sample_indices
+
+
+def is_unique(sesh, history_id, vars_id, time):
+    q = sesh.query(Obs.id).filter(
+        and_(Obs.history_id == history_id, Obs.vars_id == vars_id,
+             Obs.time == time))
+    if q.count() > 0:
+        return False
+
+    return True
```

Comment:
It's logically the same, but I'd probably say:

```
    else:
        return True
```

There are only two options, so it's maybe slightly more clear to be explicit that either one outcome or the other happens.